src/lib/state/DY.Lib.State.Database.fst

```diff
@@ -0,0 +1,413 @@
+  Some? row_opt
+
+val get_row_by_pointer :
+  #row_t:Type0 -> {|db_types row_t|} ->
+  tr:trace ->
+  prin:principal -> row_state_id:state_id{is_pointer_to_row #row_t tr prin row_state_id} ->
+  row_t
+let get_row_by_pointer #row_t #db_t tr prin row_state_id =
+  let (row_opt, _) = get_state #row_t prin row_state_id tr in
+  Some?.v row_opt
+
+#push-options "--fuel 1 --ifuel 1"
+val list_coerceP :
+  #a:Type0 ->
+  pred:(a -> prop) ->
+  l:list a{for_allP pred l} ->
+  list (x:a{pred x})
+let rec list_coerceP #a pred l =
+  for_allP_eq pred l;
+  match l with
+  | [] -> []
+  | h::t -> h::(list_coerceP pred t)
+#pop-options
+
+val db_session_invariant:
+  {|crypto_invariants|} ->
+  #row_t:Type0 -> {|db_types row_t|} ->
+  db_pred:db_predicate row_t ->
+  local_state_predicate db
+let db_session_invariant #cinvs #row_t #db_t db_pred = {
+  pred = (fun tr prin sess_id db ->
+    for_allP (is_pointer_to_row #row_t tr prin) db.rows /\
+    (
+      let tmp : list (sid:state_id{is_pointer_to_row #row_t tr prin sid}) = list_coerceP (is_pointer_to_row #row_t tr prin) db.rows in
+      let rows = List.Tot.map (get_row_by_pointer #row_t tr prin) tmp in
+      db_pred.db_global_pred tr prin sess_id rows
+    )
+  );
+  pred_later = (fun tr1 tr2 prin sess_id db ->
+    // Problem: If the db_data changes from tr1 to tr2 (via an update to one of the state IDs), then this doesn't follow from db_global_pred_later.
+    // db_global_pred_later
+    // Do we need to revise what input we have?
+    // Do we already need some monotonicity on the individual rows?
+    // Idea: monotonicity on rows (specified by some row-evolution predicate).
+    //   db_global_pred_later assumes that the rows have only updated according to row-evolution predicate, proves global pred later holds.
+    //   Proof relies on saying that the db_data updates from tr1 to tr2 according to this evolution predicate (row-wise), and then using
+    //   this to apply db_global_pred_later.
+    admit()
+  );
+  pred_knowable = (fun tr prin sess_id db ->
+    // Intuition: apply literal_to_bytes_is_publishable to each element of the list
+    admit()
+  );
+}
+
+#push-options "--fuel 1 --ifuel 1"
+val db_invariant_empty :
+  {|crypto_invariants|} ->
+  #row_t:Type0 -> {|db_types row_t|} ->
+  db_pred:db_predicate row_t ->
+  tr:trace ->
+  prin:principal -> sess_id:state_id ->
+  Lemma
+  (requires db_pred.db_global_pred tr prin sess_id [])
+  (ensures (
+   let empty_db : db = { rows = [] } in
+   (db_session_invariant db_pred).pred tr prin sess_id empty_db
+  ))
+let db_invariant_empty #cinvs #row_t #db_t db_pred tr prin sess_id =
+  let tmp = list_coerceP (is_pointer_to_row #row_t tr prin) [] in
+  ()
+#pop-options
+
+val has_db_invariants:
+  #row_t:Type0 -> {|db_types row_t|} ->
+  {|protocol_invariants|} -> db_pred:db_predicate row_t -> prop
+let has_db_invariants #row_t #db_t #invs db_pred =
+  let db_local_state : local_state db = local_state_db row_t in
```

Comment:
I've tried to avoid needing this, but for some reason, can't get typeclass inference to work property to find an instance of local_state db, probably because `db` doesn't depend on the row type. Maybe the `db_types` need to be split into two parts, one that depends on the row type and one that doesn't?